File: examples/producer/rust/src/producer/responses/userlogin_request.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use clibri::server;
use uuid::Uuid;

type BroadcastEventsUserConnected = (Vec<Uuid>, protocol::Events::UserConnected);
type BroadcastEventsMessage = (Vec<Uuid>, protocol::Events::Message);

pub enum Response {
    Accept(
        (
            protocol::UserLogin::Accepted,
            BroadcastEventsUserConnected,
            BroadcastEventsMessage,
        ),
    ),
    Deny(protocol::UserLogin::Denied),
}

#[allow(unused_variables)]
pub async fn response<E: server::Error, C: server::Control<E>>(
    request: &protocol::UserLogin::Request,
    scope: &mut Scope<'_, E, C>,
) -> Result<Response, protocol::UserLogin::Err> {
    let uuid = scope.identification.uuid();
    if scope.context.is_user_exist(&request.username).await {
        Ok(Response::Deny(protocol::UserLogin::Denied {
            reason: String::from("User has been login already"),
        }))
    } else {
        scope.context.add_user(uuid, &request.username).await;
        let msg = scope
            .context
            .add_message(
                &request.username,
                format!("User {} has been join to chat", request.username),
            )
            .map_err(|e| protocol::UserLogin::Err { error: e })?;
        Ok(Response::Accept((
            protocol::UserLogin::Accepted {
                uuid: uuid.to_string(),
            },
            (
                scope.filter.except(&uuid),
                protocol::Events::UserConnected {
                    username: request.username.clone(),
                    uuid: uuid.to_string(),
                },
            ),
            (scope.filter.except(&uuid), msg),
        )))
    }
}

File: examples/producer/typescript/src/producer/responses/userlogin.request.ts
import { Response } from "../implementation/responses/userlogin.request";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.UserLogin.Request,
	scope: Scope
): Promise<Response> {
	scope.context.addUser(scope.consumer.uuid(), request.username);
	const msg = scope.context.addMessage(
		new Protocol.Messages.Message({
			user: request.username,
			uuid: scope.consumer.uuid(),
			message: `User ${request.username} has been join to chat`,
			timestamp: BigInt(Date.now()),
		})
	);
	return Promise.resolve(
		new Response(
			new Protocol.UserLogin.Accepted({ uuid: scope.consumer.uuid() })
		)
			.broadcast(scope.filter.except(scope.consumer.uuid()))
			.EventsMessage(
				new Protocol.Events.Message({
					uuid: msg.uuid,
					user: msg.user,
					message: msg.message,
					timestamp: msg.timestamp,
				})
			)
			.broadcast(scope.filter.except(scope.consumer.uuid()))
			.EventsUserConnected(
				new Protocol.Events.UserConnected({
					username: request.username,
					uuid: scope.consumer.uuid(),
				})
			)
	);
}
